sql/mozfun/map/mode_last_struct_retain_nulls/udf.sql

```diff
@@ -0,0 +1,251 @@
+            'Munich' AS city,
+            'BY' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          ),
+          STRUCT(
+            'Berlin' AS city,
+            'BE' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          ),
+          STRUCT(
+            'Munich' AS city,
+            'BY' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          )  -- latest among the tied
+        ]
+      )
+    )
+  ),
+  -- 3) FULL-struct equality: different subdivision2 means different value
+  assert.equals(
+    TO_JSON_STRING(
+      STRUCT('Berlin' AS city, 'BE' AS subdivision1, 'A' AS subdivision2, 'DE' AS country)
+    ),
+    TO_JSON_STRING(
+      map.mode_last_struct_retain_nulls(
+        [
+          STRUCT('Berlin' AS city, 'BE' AS subdivision1, 'A' AS subdivision2, 'DE' AS country),
+          STRUCT('Berlin' AS city, 'BE' AS subdivision1, 'B' AS subdivision2, 'DE' AS country),
+          STRUCT('Berlin' AS city, 'BE' AS subdivision1, 'A' AS subdivision2, 'DE' AS country)
+        ]
+      )
+    )
+  ),
+  -- 4) Single element returns itself
+  assert.equals(
+    TO_JSON_STRING(
+      STRUCT(
+        'Cologne' AS city,
+        'NW' AS subdivision1,
+        CAST(NULL AS STRING) AS subdivision2,
+        'DE' AS country
+      )
+    ),
+    TO_JSON_STRING(
+      map.mode_last_struct_retain_nulls(
+        [
+          STRUCT(
+            'Cologne' AS city,
+            'NW' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          )
+        ]
+      )
+    )
+  ),
+-- 5) Tie between NULL struct and a non-NULL struct; latest wins -> expect Berlin
+  assert.equals(
+    TO_JSON_STRING(
+      STRUCT(
+        'Berlin' AS city,
+        'BE' AS subdivision1,
+        CAST(NULL AS STRING) AS subdivision2,
+        'DE' AS country
+      )
+    ),
+    TO_JSON_STRING(
+      map.mode_last_struct_retain_nulls(
+        [
+          STRUCT(
+            CAST(NULL AS STRING) AS city,
+            CAST(NULL AS STRING) AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            CAST(NULL AS STRING) AS country
+          ),
+          STRUCT(
+            'Berlin' AS city,
+            'BE' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          ),
+          STRUCT(
+            CAST(NULL AS STRING) AS city,
+            CAST(NULL AS STRING) AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            CAST(NULL AS STRING) AS country
+          ),
+          STRUCT(
+            'Berlin' AS city,
+            'BE' AS subdivision1,
+            CAST(NULL AS STRING) AS subdivision2,
+            'DE' AS country
+          )   -- latest among the tied
+        ]
+      )
+    )
+  ),
+-- 6) NULL struct occurs most frequently -> expect NULL
```

Comment:
This makes sense for this udf but for https://github.com/mozilla/bigquery-etl/pull/7974, wouldn't we want to consider Berlin as the first/last seen city in this case?

Reply:
I'm not sure that matches what I was thinking so I might be misunderstanding where this is going to be used.  I'll see if it makes sense when I look at where the udf is used